File: java-pi-pantry/src/main/java/com/langrsoft/pi/pantry/ItemFactory.java
package com.langrsoft.pi.pantry;

import java.time.Clock;
import java.time.LocalDate;
import java.util.HashMap;
import java.util.Map;

import com.langrsoft.util.JsonUtil;

public class ItemFactory {
    private Map<String, String> numberToLocalNameMappings = new HashMap<>();
    private Clock clock = Clock.systemDefaultZone();
    private ShelfLifeData shelfLifeData = new ShelfLifeData();

    public Item create(String json ) {
        Item item = JsonUtil.parse(json, Item.class);
        item.setSourceName(item.getName());
        item.setExpirationDate(LocalDate.MAX);
        item.setSellByDate(LocalDate.now(clock));
        changeNameIfLocalMappingExists(item);
        changeCategoryIfRecognized(item);
        return item;
    }

    private void changeCategoryIfRecognized(Item item) {
        if (shelfLifeData.contains(item.getName()))
            item.setCategory(item.getName());
    }

    private void changeNameIfLocalMappingExists(Item item) {
        if (numberToLocalNameMappings.containsKey(item.getNumber()))
            item.setName(numberToLocalNameMappings.get(item.getNumber()));
    }

    public void setNumberToLocalNameMappings(Map<String, String> numberToLocalNameMappings) {
        this.numberToLocalNameMappings = numberToLocalNameMappings;
    }

    public void setClock(Clock clock) {
        this.clock = clock;
    }
}

File: csharp-pi-pantry/csharp-pi-pantry/Domain/ItemFactory.cs
using Pipantry.Util;
using System;
using System.Collections.Generic;

namespace Pipantry.Domain
{
    public class ItemFactory
    {
        private IDictionary<string, string> numberToLocalNameMappings = new Dictionary<string, string>();
        private ShelfLifeRepository shelfLifeData = new ShelfLifeRepository();

        public ItemFactory()
        {
            DateTimeProvider = new DateTimeWrapper();
        }

        public IDateTime DateTimeProvider { get; set; }

        public Item Create(string json)
        {
            var item = JsonUtil.Parse<Item>(json);
            item.SourceName = item.Name;
            item.ExpirationDate = DateTime.MaxValue;
            item.SellByDate = DateTimeProvider.Now;
            ChangeNameIfLocalMappingExists(item);
            ChangeCategoryIfRecognized(item);
            return item;
        }

        private void ChangeCategoryIfRecognized(Item item)
        {
            if (item.Name != null && shelfLifeData.ContainsKey(item.Name))
                item.Category = item.Name;
        }

        private void ChangeNameIfLocalMappingExists(Item item)
        {
            if (item.Number != null && numberToLocalNameMappings.ContainsKey(item.Number))
                item.Name = numberToLocalNameMappings[item.Number];
        }

        public void setNumberToLocalNameMappings(IDictionary<string, string> numberToLocalNameMappings)
        {
            this.numberToLocalNameMappings = numberToLocalNameMappings;
        }
    }
}
